### PR TITLE
Snode version patch

### PR DESCRIPTION
**Pull request by ThomasSession**
We are not fetching snode versions and using that to check the last node in an onion request 
to make sure its version is >= 2.8.0

### CODE REVIEW

**Review by mpretty-cyro — APPROVED** [review]

File: libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt [hunk 1]
```diff
@@ -190,8 +191,19 @@ object OnionRequestAPI {
...
+                        var pathSnode = unusedSnodes.getRandomElement()
+
+                        // we want to make sure the last node in the path is above version 2.8.0
+                        // to help with an issue that will disappear once the nodes are all updated
+                        if(index == pathSize - 2) {
+                            // because we are now grabbing the whole node pool there should always
+                            // be a node that is above version 2.8.0
+                            while(Util.compareVersions(pathSnode.version, "2.8.0") < 0) {
+                                pathSnode = unusedSnodes.getRandomElement()
+                            }
```
Comment (simophin) [inline libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt:204]: Is the assumption "should always be a node that is above version 2.8.0" reliable? I'm a little bit worried that this can create a infinite loop in exceptional cases, if the server side ends up breaking the assumption. Maybe remove the node from the comparison pool and check if the pool still has something as a safe guard?

File: libsession/src/main/java/org/session/libsession/snode/SnodeAPI.kt [hunk 2]
```diff
@@ -206,6 +226,10 @@ object SnodeAPI {
         }
     }
 
+    private fun extractVersionString(jsonVersion: String): String{
+        return jsonVersion.removeSurrounding("[", "]").split(", ").joinToString(separator = ".")
```
Comment (simophin) [inline libsession/src/main/java/org/session/libsession/snode/SnodeAPI.kt:230]: Nice! I did not know `removeSurrouding` exists, I'd almost grab regex to start with

File: libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt [hunk 3]
```diff
@@ -190,8 +191,19 @@ object OnionRequestAPI {
...
                 guardSnodes.minus(reusableGuardSnodes).map { guardSnode ->
-                    val result = listOf( guardSnode ) + (0 until (pathSize - 1)).map {
-                        val pathSnode = unusedSnodes.getRandomElement()
+                    val result = listOf( guardSnode ) + (0 until (pathSize - 1)).mapIndexed() { index, _ ->
+                        var pathSnode = unusedSnodes.getRandomElement()
+
+                        // we want to make sure the last node in the path is above version 2.8.0
+                        // to help with an issue that will disappear once the nodes are all updated
+                        if(index == pathSize - 2) {
+                            // because we are now grabbing the whole node pool there should always
```
Comment (AL-Session) [inline libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt:200]: I'd rephrase this comment to "Keep picking a random unused node until we find one that is at least version 2.8.0.."

File: libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt (hunk 1 above)
Comment (simophin) [reply libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt:204]: This may be one of the implementations?
```suggestion
  pathSnode = unusedSnodes.shuffled().firstOrNull { Util.compareVersions(pathSnode.version, "2.8.0") >= 0 }
```

File: libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt [hunk 4]
```diff
@@ -190,8 +191,19 @@ object OnionRequestAPI {
...
+                            // be a node that is above version 2.8.0
+                            while(Util.compareVersions(pathSnode.version, "2.8.0") < 0) {
+                                pathSnode = unusedSnodes.getRandomElement()
+                            }
+                        }
+
                         unusedSnodes = unusedSnodes.minus(pathSnode)
```
Comment (AL-Session) [inline libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt:207]: "..and then remove that node from the list of unused nodes and return it".

File: libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt [hunk 5]
```diff
@@ -190,8 +191,19 @@ object OnionRequestAPI {
...
                 // Don't test path snodes as this would reveal the user's IP to them
                 guardSnodes.minus(reusableGuardSnodes).map { guardSnode ->
-                    val result = listOf( guardSnode ) + (0 until (pathSize - 1)).map {
-                        val pathSnode = unusedSnodes.getRandomElement()
+                    val result = listOf( guardSnode ) + (0 until (pathSize - 1)).mapIndexed() { index, _ ->
+                        var pathSnode = unusedSnodes.getRandomElement()
+
+                        // we want to make sure the last node in the path is above version 2.8.0
+                        // to help with an issue that will disappear once the nodes are all updated
+                        if(index == pathSize - 2) {
```
Comment (AL-Session) [inline libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt:199]: Why `pathsize - 2`?

File: libsession/src/main/java/org/session/libsession/snode/SnodeAPI.kt [hunk 6]
```diff
@@ -173,12 +183,22 @@ object SnodeAPI {
...
+                            val address = rawSnodeAsJSON?.get(KEY_IP) as? String
+                            val port = rawSnodeAsJSON?.get(KEY_PORT) as? Int
+                            val ed25519Key = rawSnodeAsJSON?.get(KEY_ED25519) as? String
+                            val x25519Key = rawSnodeAsJSON?.get(KEY_X25519) as? String
+                            val version = (rawSnodeAsJSON?.get(KEY_VERSION) as? ArrayList<*>)
+                                ?.filterIsInstance<Int>() // get the array as Integers
+                                ?.joinToString(separator = ".") // turn it int a version string
+
+                            if (address != null && port != null && ed25519Key != null && x25519Key != null
+                                && address != "0.0.0.0" && version != null) {
```
Comment (AL-Session) [inline libsession/src/main/java/org/session/libsession/snode/SnodeAPI.kt:195]: Could move address != "0.0.0.0" closer to the null check for it if you wanted, or not.

File: app/src/main/java/org/thoughtcrime/securesms/database/LokiAPIDatabase.kt [hunk 7]
```diff
@@ -179,7 +179,8 @@ class LokiAPIDatabase(context: Context, helper: SQLCipherOpenHelper) : Database(
                 val port = components.getOrNull(1)?.toIntOrNull() ?: return@mapNotNull null
                 val ed25519Key = components.getOrNull(2) ?: return@mapNotNull null
                 val x25519Key = components.getOrNull(3) ?: return@mapNotNull null
-                Snode(address, port, Snode.KeySet(ed25519Key, x25519Key))
+                val version = components.getOrNull(4) ?: "0.0.0"
```
Comment (AL-Session) [inline app/src/main/java/org/thoughtcrime/securesms/database/LokiAPIDatabase.kt:182]: We use this "0.0.0" for unknown / null version number a bit - could make it into a constant to avoid fat-finger errors.

**Review by AL-Session — APPROVED** [review]
LGTM - couple of minor suggestions but I'm happy if you are. The only part I don't understand is where `pathSize - 2` is coming from.

File: libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt [hunk 8]
```diff
@@ -190,8 +191,19 @@ object OnionRequestAPI {
...
+                            // because we are now grabbing the whole node pool there should always
+                            // be a node that is above version 2.8.0
+                            while(Util.compareVersions(pathSnode.version, "2.8.0") < 0) {
```
Comment (AL-Session) [inline libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt:202]: One final thing, sometimes I'll put `emergencyExit` conditions on while loops like this just in case for whatever god-forsaken reason we end up trapped. Like if we go through the loop 1000 times then nah (maybe we have a small pool size, maybe the RNG is giving us the same node over and over, maybe anything..). Up to you, though.

Comment (AL-Session) [inline libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt:202]: If you hard-code the "0.0.0" into a constant maybe do this "2.8.0" as `MINIMUM_EXIT_NODE_VERSION` as well.

File: libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt (hunk 5 above)
Comment (ThomasSession) [reply libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt:199]: `until` is non inclusive. This is because the first node is actually captured with `guardSnode` 
` listOf( guardSnode ) +`

File: libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt (hunk 8 above)
Comment (ThomasSession) [reply libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt:202]: These will be removed in the next release, so this is all temporary

File: libsession/src/main/java/org/session/libsession/snode/SnodeAPI.kt [hunk 9]
```diff
@@ -716,10 +740,11 @@ object SnodeAPI {
...
                 val portAsString = rawSnodeAsJSON?.get("port") as? String
                 val port = portAsString?.toInt()
-                val ed25519Key = rawSnodeAsJSON?.get("pubkey_ed25519") as? String
-                val x25519Key = rawSnodeAsJSON?.get("pubkey_x25519") as? String
+                val ed25519Key = rawSnodeAsJSON?.get(KEY_ED25519) as? String
+                val x25519Key = rawSnodeAsJSON?.get(KEY_X25519) as? String
+
                 if (address != null && port != null && ed25519Key != null && x25519Key != null && address != "0.0.0.0") {
-                    Snode("https://$address", port, Snode.KeySet(ed25519Key, x25519Key))
+                    Snode("https://$address", port, Snode.KeySet(ed25519Key, x25519Key), "0.0.0")
```
Comment (bemusementpark) [inline libsession/src/main/java/org/session/libsession/snode/SnodeAPI.kt:747]: lol, I saw `0.0.0` and thought malformed IP address, the const would be useful.

File: libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt (hunk 1 above)
Comment (ThomasSession) [reply libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt:204]: I was wondering the same but Session controls about 10% of the nodes and they are guaranteed to be updated.
This patch will also be removed on the next release as all nodes will have been updated by then.

File: libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt (hunk 5 above)
Comment (AL-Session) [reply libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt:199]: Okey doke!

File: libsession/src/main/java/org/session/libsession/utilities/Util.kt [hunk 10]
```diff
@@ -365,6 +365,34 @@ object Util {
...
+        val paddedParts2 = parts2 + List(maxLength - parts2.size) { 0 }
+
+        for (i in 0 until maxLength) {
+            val compare = paddedParts1[i].compareTo(paddedParts2[i])
+            if (compare != 0) {
+                return compare
+            }
+        }
+        return 0
+    }
```
Comment (bemusementpark) [inline libsession/src/main/java/org/session/libsession/utilities/Util.kt:395]: Can maybe just use `List.compareTo` ?

```suggestion
    fun compareVersions(version1: String, version2: String): Int {
        val parts1 = version1.split(".").map { it.toIntOrNull() ?: 0 }
        val parts2 = version2.split(".").map { it.toIntOrNull() ?: 0 }

        val maxLength = maxOf(parts1.size, parts2.size)
        val paddedParts1 = parts1 + List(maxLength - parts1.size) { 0 }
        val paddedParts2 = parts2 + List(maxLength - parts2.size) { 0 }

        return paddedParts1.compareTo(paddedParts2)
    }
```

otherwise you don't need to construct the lists:

```suggestion
    fun compareVersions(version1: String, version2: String): Int {
        val parts1 = version1.split(".").map { it. toIntOrNull() }
        val parts2 = version2.split(".").map { it. toIntOrNull() }

        val maxLength = maxOf(parts1.size, parts2.size)

        for (i in 0 until maxLength) {
            val v1 = parts1.getOrElse(i) { 0 }
            val v2 = parts2.getOrElse(i) { 0 }
            if (v1 != v2) return v1.compareTo(v2)
        }
        return 0
}
```

File: app/src/main/java/org/thoughtcrime/securesms/ApplicationContext.java [hunk 11]
```diff
@@ -214,6 +214,17 @@ public void onCreate() {
         DatabaseModule.init(this);
         MessagingModuleConfiguration.configure(this);
         super.onCreate();
+
+        // we need to clear the snode and onionrequest databases once on first launch
+        // in order to apply a patch that adds a version number to the Snode objects.
+        if(!TextSecurePreferences.hasAppliedPatchSnodeVersion(this)) {
+            ThreadUtils.queue(() -> {
```
[the remark this answers is not in the document]
Comment (ThomasSession) [reply app/src/main/java/org/thoughtcrime/securesms/ApplicationContext.java:221]: We can't from a Java class I think

File: libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt [hunk 12]
```diff
@@ -190,8 +191,19 @@ object OnionRequestAPI {
                 if (unusedSnodes.count() < pathSnodeCount) { throw InsufficientSnodesException() }
                 // Don't test path snodes as this would reveal the user's IP to them
                 guardSnodes.minus(reusableGuardSnodes).map { guardSnode ->
-                    val result = listOf( guardSnode ) + (0 until (pathSize - 1)).map {
-                        val pathSnode = unusedSnodes.getRandomElement()
+                    val result = listOf( guardSnode ) + (0 until (pathSize - 1)).mapIndexed() { index, _ ->
```
[the remark this answers is not in the document]
Comment (ThomasSession) [reply libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt:194]: The range on this is wrong

File: libsession/src/main/java/org/session/libsession/utilities/Util.kt (hunk 10 above)
Comment (ThomasSession) [reply libsession/src/main/java/org/session/libsession/utilities/Util.kt:395]: Neither of these two options work if you test them out?

File: libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt [hunk 13]
```diff
@@ -190,8 +191,21 @@ object OnionRequestAPI {
...
+                        // For the last node: We need to make sure the version is >= 2.8.0
+                        // to help with an issue that will disappear once the nodes are all updated
+                        if(index == pathSize - 2) {
+                            val suitableSnodes = unusedSnodes.filter { Util.compareVersions(it.version, "2.8.0") >= 0 }
+                            pathSnode = if (suitableSnodes.isNotEmpty()) {
+                                suitableSnodes.random()
+                            } else {
+                                throw InsufficientSnodesException()
+                            }
+                        }
```
Comment (bemusementpark) [inline libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt:206]: Maybe something like this is (marginally more efficient and) resolves the `wtf` you get seeing the `pathSize - 2`

```suggestion
                    val result = buildList {
                        add(guardSnode)
                        addAll((1 until (pathSize)).mapIndexed { index ->
                            var pathSnode = unusedSnodes.getRandomElement()

                            // For the last node: We need to make sure the version is >= 2.8.0
                            // to help with an issue that will disappear once the nodes are all updated
                            if(index == pathSize - 1) {
                                val suitableSnodes = unusedSnodes.filter { Util.compareVersions(it.version, "2.8.0") >= 0 }
                                pathSnode = if (suitableSnodes.isNotEmpty()) {
                                    suitableSnodes.random()
                                } else {
                                    throw InsufficientSnodesException()
                                }
                            }
                        )
```

File: libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (bemusementpark) [reply libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt:194]: yeah...

**Review by simophin — APPROVED** [review]

File: libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt (hunk 13 above)
Comment (ThomasSession) [reply libsession/src/main/java/org/session/libsession/snode/OnionRequestAPI.kt:206]: I'm not too worried as this is to be removed on the next release.
